addons/website_sale_coupon/models/sale_order.py

```diff
@@ -8,6 +8,26 @@
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
+    def _try_pending_coupon(self):
+        if not request:
+            return
+
+        status = {key: value for key, value in request.session.get('pending_coupon', {}).items() if key == 'promo_code'}
+
+        if self and 'promo_code' in status:
+            coupon_code = status.get('promo_code')
+            error_status = self.env['sale.coupon.apply.code'].apply_coupon(self, coupon_code)
+            status.update(error_status)
+            if coupon_code == self.code_promo_program_id.promo_code or coupon_code in self.applied_coupon_ids.mapped('code'):
+                status.pop('promo_code')
+
+        request.session['pending_coupon'] = status
+
+    def recompute_coupon_lines(self):
+        result = super().recompute_coupon_lines()
+        self._try_pending_coupon()
```

Comment:
Isn't it the other way arround? trypending then recompute?
Imagine you have a cart like that:
- 3 ipads 3x300€
- 20% on cart so 180€ discount
Now you apply the coupon you received that add a free ipad
- you will have 3 ipads 3x300€
- one free ipd 300€
- still 180€ discount but it should be 120€ since you paid only 2 ipads ?

To check.. but pretty sure its the case.. 

If what I describe here is right and it need to be fixed, can you write a test about it? Eg a test that create such a cart then create a share coupon link and apply it -> check everything is fine in cart